### PR TITLE
Incremental loading

### PR DESCRIPTION
**Pull request by SarahRo**
This PR implements the following non-linear solution strategy for contact problem:
Denote by $u$ a finite element function, by $v$ a test function and by $x$ the spatial coordinates, We write the nonliner variational form as
$0 = F(u, v, x) + G(v, x)$, 
where $G(x)$ refers to some given data such as boundary conditions or body forces. Instead of solving the non-linear equation directly, we solve $n$ non-linear equations for $\Delta u_i$
$0 = F(u_i, v, x) + \lambda_i G(v,x) = F(u_{i-1} + \Delta u_i, v, x) + \lambda_i G(v,x)$, 
where $u_0 = 0$, $\lambda_i \in [0, 1]$, $\lambda_i> \lambda_{i-1}$, $\lambda_n = 1$.  
The Newton iteration in this case takes the following form:
$\Delta u_i^{k+1} = \Delta u_i^k - (D_{\Delta u_i} F(u_{i-1} + \Delta u_i, v, x))^{-1}(F(u_{i-1} + \Delta u_i, v, x) + \lambda_i G(v, x))$
So far we have assumed that $F$ is independent of $\lambda_i$. This does not hold if the gap function is recomputed for each $\lambda_i$. This will affect the performance of the solution algorithm if the solution of the employed projection strategy varies significantly from one step to the next as the initial guess $\Delta u_i^0 = 0$ will not be as accurate.

### CODE REVIEW

File: python/dolfinx_contact/unbiased/nitsche_unbiased.py [hunk 1]
```diff
@@ -22,22 +21,219 @@
...
+    ==========
+    incr               The load increment in [0,1]
+    F_custom           The compiled form for the residual vector
+    J_custom           The compiled form for the jacobian
+    u                  The displacement from previous step
+    du                 The change in displacement to be computed
+    contact            The contact class
+    markers            The meshtags marking surfaces and domains
+    entities           The contact surface entities for integration
+    quadrature_degre   The quadrature degree
```
Comment (jorgensd) [inline python/dolfinx_contact/unbiased/nitsche_unbiased.py:44]:
```suggestion
    quadrature_degree   The quadrature degree
```

File: python/dolfinx_contact/unbiased/nitsche_unbiased.py [hunk 2]
```diff
@@ -22,22 +21,219 @@
...
+
+    # pack grad u
+    grad_u = []
+    with common.Timer("~~Contact: Pack u"):
+        for i in range(num_pairs):
+            grad_u.append(dolfinx_contact.cpp.pack_gradient_quadrature(
+                u._cpp_object, quadrature_degree, entities[i]))
+
+    # function for updating coefficients coefficients
+    @ common.timed("~Contact: Update coefficients")
```
Comment (jorgensd) [inline python/dolfinx_contact/unbiased/nitsche_unbiased.py:98]:
```suggestion
    @common.timed("~Contact: Update coefficients")
```

File: python/dolfinx_contact/unbiased/nitsche_unbiased.py [hunk 3]
```diff
@@ -22,22 +21,219 @@
...
+    def compute_coefficients(x, coeffs):
+        size_local = V.dofmap.index_map.size_local
+        bs = V.dofmap.index_map_bs
+        du.x.array[:size_local * bs] = x.array_r[:size_local * bs]
+        du.x.scatter_forward()
```
Comment (jorgensd) [inline python/dolfinx_contact/unbiased/nitsche_unbiased.py:103]: What is the reason for only updating the local bit and then scatter forward, is x not been distributed, or is it because x is a petsc vector?

File: python/dolfinx_contact/unbiased/nitsche_unbiased.py [hunk 4]
```diff
@@ -22,22 +21,219 @@
...
+                u_puppet.append(dolfinx_contact.cpp.pack_coefficient_quadrature(
+                    du._cpp_object, quadrature_degree, entities[i]))
+                grad_u_puppet.append(dolfinx_contact.cpp.pack_gradient_quadrature(
+                    du._cpp_object, quadrature_degree, entities[i]))
+        for i in range(num_pairs):
+            c_0 = np.hstack([ccfs[i], u_puppet[i], grad_u_puppet[i] + grad_u[i], u_candidate[i]])
+            coeffs[i][:, :] = c_0[:, :]
+
+    # function for computing residual
+    @ common.timed("~Contact: Assemble residual")
```
Comment (jorgensd) [inline python/dolfinx_contact/unbiased/nitsche_unbiased.py:121]:
```suggestion
    @common.timed("~Contact: Assemble residual")
```

**Review by jorgensd — CHANGES_REQUESTED** [review]
Only minor stuff

File: python/dolfinx_contact/unbiased/nitsche_unbiased.py (hunk 3 above)
Comment (SarahRo) [reply python/dolfinx_contact/unbiased/nitsche_unbiased.py:103]: x is a petsc vector I believe

File: python/dolfinx_contact/unbiased/nitsche_unbiased.py [hunk 5]
```diff
@@ -22,22 +21,219 @@
...
+    def compute_residual(x, b, coeffs):
+        b.zeroEntries()
+        b.ghostUpdate(addv=_PETSc.InsertMode.INSERT, mode=_PETSc.ScatterMode.FORWARD)
```
[the remark this answers is not in the document]
Comment (SarahRo) [reply python/dolfinx_contact/unbiased/nitsche_unbiased.py:124]: I'm pretty sure I added this because it was causing my code to fail.

File: python/dolfinx_contact/unbiased/nitsche_unbiased.py [hunk 6]
```diff
@@ -22,22 +21,219 @@
...
+    newton_solver.set_coefficients(compute_coefficients)
+
+    # Set rigid motion nullspace
+    null_space = rigid_motions_nullspace_subdomains(V, markers[0], np.unique(markers[0].values))
+    newton_solver.A.setNearNullSpace(null_space)
+
+    return newton_solver
+
+
+def get_problem_parameters(problem_parameters):
```
[the remark this answers is not in the document]
Comment (SarahRo) [reply python/dolfinx_contact/unbiased/nitsche_unbiased.py:166]: Generally it would make sense, but things are a bit awkward due to the return type of .get
